### PR TITLE
readd some of the documentation that helps to understand how RestrictedPython works internally

### PR DESCRIPTION
**Pull request by loechel**
Some of the documentation got removed, when cutting the 4.0 release.

I readd some of the documentation as a remainder how RestrictedPython works internally and the AST Grammar files as they help to understand changes between Python Versions that may affects RestrictedPython.

### CODE REVIEW

File: docs/contributing/index.rst [hunk 1]
```diff
@@ -1,9 +1,114 @@
...
+* In detail Documentation on the Python AST module ``ast`` (Green Tree Snakes) (https://greentreesnakes.readthedocs.org/en/latest/)
+* Example how to Instrumenting the Python AST (``ast.AST``) (http://www.dalkescientific.com/writings/diary/archive/2010/02/22/instrumenting_the_ast.html)
+
+.. _`CamelCase`: https://en.wikipedia.org/wiki/Camel_case
+
+.. _`EBNF`: https://en.wikipedia.org/wiki/Extended_Backus%E2%80%93Naur_form
+
+.. _`Chomsky 3`: https://en.wikipedia.org/wiki/Chomsky_hierarchy#Type-3_grammars
+
+.. _`Python 3.8 AST`: https://docs.python.org/3.7/library/ast.html#abstract-grammar
```
Comment (icemac) [inline docs/contributing/index.rst]: This links to 3.7 instead of 3.8.

File: docs/contributing/index.rst [hunk 2]
```diff
@@ -1,9 +1,114 @@
...
+    * `Python 3.7 AST`_
+    * `Python 3.6 AST`_
+    * `Python 3.5 AST`_
+    * `Python 3.4 AST`_ (obsolete)
+    * `Python 3.3 AST`_ (obsolete)
+    * `Python 3.2 AST`_ (obsolete)
+    * `Python 3.1 AST`_ (obsolete)
+    * `Python 3.0 AST`_ (obsolete)
+    * `Python 2.7 AST`_
+    * `Python 2.6 AST`_ (obsolete)
```
Comment (icemac) [inline docs/contributing/index.rst]: I am not sure if it helps to list so many obsolete Python grammars.

File: docs/contributing/index.rst [hunk 3]
```diff
@@ -1,9 +1,114 @@
...
+
+
+``NodeVisitor`` & ``NodeTransformer``
++++++++++++++++++++++++++++++++++++++
+
+A ``NodeVisitor`` is a class of a node / AST consumer, it reads the data by stepping through the tree without modifying it.
+In contrast, a ``NodeTransformer`` (which inherits from a ``NodeVisitor``) is allowed to modify the tree and nodes.
+
+Modifying the AST
++++++++++++++++++
```
Comment (icemac) [inline docs/contributing/index.rst]: This headline has no section content, why?

File: docs/contributing/index.rst [hunk 4]
```diff
@@ -1,9 +1,114 @@
...
+    * `Python 3.0 AST`_ (obsolete)
+    * `Python 2.7 AST`_
+    * `Python 2.6 AST`_ (obsolete)
+
+  * ``NodeVistiors``  (https://docs.python.org/3.5/library/ast.html#ast.NodeVisitor)
+  * ``NodeTransformer``  (https://docs.python.org/3.5/library/ast.html#ast.NodeTransformer)
+  * ``dump`` (https://docs.python.org/3.5/library/ast.html#ast.dump)
+
+* In detail Documentation on the Python AST module ``ast`` (Green Tree Snakes) (https://greentreesnakes.readthedocs.org/en/latest/)
+* Example how to Instrumenting the Python AST (``ast.AST``) (http://www.dalkescientific.com/writings/diary/archive/2010/02/22/instrumenting_the_ast.html)
```
Comment (icemac) [inline docs/contributing/index.rst]: This could also be a link, and the one in the line above, too.

File: docs/contributing/index.rst (hunk 1 above)
Comment (loechel) [reply docs/contributing/index.rst]: fixed

File: docs/contributing/index.rst (hunk 2 above)
Comment (loechel) [reply docs/contributing/index.rst]: with the current supported / used versions that helps, we can later remove some, which we don't need anymore, when we drop finally support for Python 2

File: docs/contributing/index.rst (hunk 3 above)
Comment (loechel) [reply docs/contributing/index.rst]: removed

File: docs/contributing/index.rst (hunk 4 above)
Comment (loechel) [reply docs/contributing/index.rst]: the problem is that ``code`` are not allowed in links, but readability counts more than styling.  changed all to link

**Comment by loechel** [comment]
I added more rst files that shows the diff between the ast files. 

This Diff is some of the bases for looking into changes and updates.

File: docs/contributing/index.rst (hunk 2 above)
Comment (icemac) [reply docs/contributing/index.rst]: okay

File: docs/contributing/changes_from37to38.rst [hunk 5]
```diff
@@ -0,0 +1,5 @@
+Changes from Python 3.7 to Python 3.8
+-------------------------------------
+
+.. literalinclude:: ast/python3_8.ast
+   :diff: ast/python3_7.ast
```
Comment (icemac) [inline docs/contributing/changes_from37to38.rst]: This is a good use case for the grammar files. I did not even know this was possible.

**Pull request merged**